File: gmail/src/gmail.ts
import { google } from 'googleapis';
import { getAuthenticatedClient } from './googleAuth';

export interface EmailFull {
  id: string;
  date: string;
  from: string;
  subject: string;
  snippet: string;
  body: string;
}

function getHeader(headers: any[], name: string): string {
  return headers.find((h: any) => h.name.toLowerCase() === name.toLowerCase())?.value ?? '';
}

function findPart(payload: any, mimeType: string): string | null {
  if (!payload) return null;
  if (payload.mimeType === mimeType && payload.body?.data) {
    return Buffer.from(payload.body.data, 'base64').toString('utf-8');
  }
  for (const part of payload.parts ?? []) {
    const found = findPart(part, mimeType);
    if (found) return found;
  }
  return null;
}

function extractBody(payload: any): string {
  let body = findPart(payload, 'text/plain');
  if (!body) {
    const html = findPart(payload, 'text/html');
    if (html) body = html.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim();
  }
  return body ?? '(no body)';
}

export async function getEmailsFull(options: {
  maxResults?: number;
  daysBack?: number;
} = {}): Promise<EmailFull[]> {
  const auth = getAuthenticatedClient();
  const gmail = google.gmail({ version: 'v1', auth });

  let q = '';
  if (options.daysBack) {
    const d = new Date();
    d.setDate(d.getDate() - options.daysBack);
    q = `after:${d.toISOString().split('T')[0].replace(/-/g, '/')}`;
  }

  const listRes = await gmail.users.messages.list({
    userId: 'me',
    q: q || undefined,
    maxResults: options.maxResults ?? 20,
  });

  const messages = listRes.data.messages ?? [];
  if (messages.length === 0) return [];

  // Fetch all full messages in parallel
  const results = await Promise.all(
    messages.map(async ({ id }) => {
      const { data } = await gmail.users.messages.get({
        userId: 'me',
        id: id!,
        format: 'full',
      });
      const headers = data.payload?.headers ?? [];
      return {
        id: id!,
        date: getHeader(headers, 'Date'),
        from: getHeader(headers, 'From'),
        subject: getHeader(headers, 'Subject'),
        snippet: data.snippet ?? '',
        body: extractBody(data.payload),
      };
    })
  );

  return results;
}

// Still exported for use in test.ts
export async function getEmailBody(id: string): Promise<EmailFull> {
  const auth = getAuthenticatedClient();
  const gmail = google.gmail({ version: 'v1', auth });

  const { data } = await gmail.users.messages.get({
    userId: 'me',
    id,
    format: 'full',
  });

  const headers = data.payload?.headers ?? [];
  return {
    id,
    date: getHeader(headers, 'Date'),
    from: getHeader(headers, 'From'),
    subject: getHeader(headers, 'Subject'),
    snippet: data.snippet ?? '',
    body: extractBody(data.payload),
  };
}

// Kept for test.ts compatibility
export async function getEmails(options: {
  maxResults?: number;
  daysBack?: number;
} = {}): Promise<EmailFull[]> {
  return getEmailsFull(options);
}

File: get-video/app/gmail.py
import base64
import logging
import re
from datetime import datetime, timedelta, timezone

from googleapiclient.discovery import build

from .google_auth import get_credentials
from .models import Email

log = logging.getLogger(__name__)


def _get_header(headers: list, name: str) -> str:
    for h in headers:
        if h.get("name", "").lower() == name.lower():
            return h.get("value", "")
    return ""


def _find_part(payload: dict, mime_type: str) -> str | None:
    if not payload:
        return None
    if payload.get("mimeType") == mime_type:
        data = payload.get("body", {}).get("data")
        if data:
            return base64.urlsafe_b64decode(data + "==").decode("utf-8", errors="replace")
    for part in payload.get("parts", []):
        found = _find_part(part, mime_type)
        if found:
            return found
    return None


def _extract_body(payload: dict) -> str:
    body = _find_part(payload, "text/plain")
    if not body:
        html = _find_part(payload, "text/html")
        if html:
            body = re.sub(r"<[^>]+>", " ", html)
            body = re.sub(r"\s+", " ", body).strip()
    return body or "(no body)"


def _fetch_message(msg_id: str) -> dict:
    creds = get_credentials()
    svc = build("gmail", "v1", credentials=creds)
    return svc.users().messages().get(
        userId="me", id=msg_id, format="full"
    ).execute()


def get_emails(days_back: int, max_results: int) -> list[Email]:
    log.info("  gmail: getting credentials...")
    creds = get_credentials()
    log.info("  gmail: building service...")
    service = build("gmail", "v1", credentials=creds, cache_discovery=False)
    log.info("  gmail: service ready, fetching messages...")

    q = ""
    if days_back:
        since = (datetime.now(tz=timezone.utc) - timedelta(days=days_back)).strftime("%Y/%m/%d")
        q = f"after:{since}"

    list_res = service.users().messages().list(
        userId="me",
        q=q or None,
        maxResults=max_results,
    ).execute()

    messages = list_res.get("messages", [])
    if not messages:
        return []

    full_messages = [_fetch_message(m["id"]) for m in messages]
    log.info(f"  gmail: fetched {len(full_messages)} full messages")

    emails = []
    for data in full_messages:
        headers = data.get("payload", {}).get("headers", [])
        emails.append(Email(
            id=data["id"],
            date=_get_header(headers, "Date"),
            **{"from": _get_header(headers, "From")},
            subject=_get_header(headers, "Subject"),
            snippet=data.get("snippet", ""),
            body=_extract_body(data.get("payload", {})),
        ))

    return emails
